### PR TITLE
Support Python 3.9, drop Python 3.6 support, update requirements

### LINKED ISSUE #303: Support Python 3.9 and drop support of python 3.6
Python 3.9 was released more than 6 months ago, so we should make hal-cgp compatible with it. This probably doesn't require much, simply adding it to the CI might be enough.

### PR DESCRIPTION
**Pull request by mschmidt87**
This PR makes hal-cgp compatible with Python 3.9 and removes support for python 3.6. 

To that end, several dependencies are updated, most importantly numpy  to version 1.20 which supports python 3.9. That led to a cascade of errors signaled by mypy which are fixed in this PR:
- many are mismatches between return types and variable types
- numpy now supports type hints internally. This leads to a problem with our type hints which specify a dtype for arrays. As far as I understand, before, mypy just did not check those at all. Now, mypy checks them, but numpy does not support hints for dtypes yet, so this threw an error. I fixed it by removing all dtypes from the array type ints.
- mypy now correctly complained about the type hint in the high-level api that defines e.g. `max_generations` as an integer but then sets its default value to `np.inf` which is a float. I solved this by setting its default value to the highest representable integer value. This works, but is a bit ugly. If you have a better idea, please comment.

Note: Github currently reports the checks to be in process because it's still waiting for the python 3.6 check which is no longer done in this PR. I will change this after getting approval from the reviewers to be able to merge the PR then.
Closes #303  and #314

### CODE REVIEW

File: cgp/hl_api.py [hunk 1]
```diff
@@ -73,15 +74,18 @@ def evolve(
...
+            " value smaller than the largest representable integer."
+        )
 
     ea.initialize_fitness_parents(pop, objective)
     if callback is not None:
         callback(pop)
 
     # perform evolution
-    max_fitness = np.finfo(np.float).min
+    max_fitness = np.finfo(np.float_).min
```
Comment (HenrikMettler) [inline cgp/hl_api.py:88]: Probably just a minor detail, but the DepreceationWarning suggested to use `np.float64` or `float` here, any specific reason why `np.float_`

File: cgp/ea/mu_plus_lambda.py [hunk 2]
```diff
@@ -154,7 +154,7 @@ def step(
 
         if self.local_search is not None:
             assert isinstance(pop.champion.fitness, float)
-            prev_avg_fitness: float = np.mean([ind.fitness for ind in combined])
+            prev_avg_fitness = np.mean([ind.fitness for ind in combined])
```
Comment (HenrikMettler) [inline cgp/ea/mu_plus_lambda.py:157]: Why did you have to drop the type hints here and below?

File: cgp/genome.py [hunk 3]
```diff
@@ -599,7 +599,7 @@ def _select_gene_indices_for_mutation(
 
         Returns
         ----------
-        selected_gene_indices: np.ndarray
+        np.ndarray
```
Comment (HenrikMettler) [inline cgp/genome.py:602]: Thx for fixing, that was my mistake i think

**Review by HenrikMettler — COMMENTED** [review]
Thanks a lot for doing this :) Looks godd, just a few comments before merging

**Comment by HenrikMettler** [comment]
Does this also address #314 ?

File: cgp/hl_api.py (hunk 1 above)
Comment (mschmidt87) [reply cgp/hl_api.py:88]: Ah I see, I didn't see that DeprecationWarning. Okay, I will use `float` instead. In fact, `np.float_` is just an alias for `np.float64`. Thanks.

**Comment by mschmidt87** [comment]
> Does this also address #314 ?

Actually yes, I believe so.

**Review by jakobj — APPROVED** [review]
great thanks for taking care of this! :+1:

some changes seem a bit unrelated to the PR title, maybe this can be easily solved by adapting the title? also please rewrite history to make different changes explicit if not already done.

**Pull request merged**